New compiler for release notes

sf1919 opened the pull request:
**Description of work.**
This PR covers most of the work needed to automatically compile seperate release notes during sprints. It does this by using the new amalgamate sphinx directive. As this pulls direct from the necessary folders the placeholder files for compiling in a previous iteration of the project design can be removed.

**To test:**

- Open `/docs/source/release/v6.4.0/diffraction.rst`
- Add the following script to the bottom of the page
```
Testing
#######
Engineering Diffraction
-----------------------

New features
############
.. amalgamate:: /Diffraction/Engineering/New_features

```
- Go to the folder `/docs/source/release/v6.4.0/Diffraction/Engineering/New_features`
- Add two or more .rst files containing text (as if they were two new separate release notes
- delete your` build/docs` folder
- run CMake and rebuild the docs
- open `build/docs/html/release/v6.4.0/diffraction.html`
- At the bottom of the page should be the new text you added to the bottom of `diffraction,rst` and the text from the two (or more) files you created under the New Features heading.

Part of #33593 


*This does not require release notes* because **it should not affect users**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
New compiler for release notes

sf1919 committed:
Add new amalgamate directive

- a new sphinx directive that compiles and displays contents of .rst files from nominated folder
- updated conf file to include amalgamate directive

sf1919 committed:
Removed unneccessary files

- blank compilation files no longer needed when using the new amalgamate directive

sf1919 committed:
Update documentation

- Update Release Note Guidance to reflect changes in how automated compilation works

sf1919 committed:
Separate compilation script no longer required.

- As notes will be compiled via a new sphinx directive a separate compilation script is no longer required, even for release.

DavidFair commented on dev-docs/source/Standards/ReleaseNotesGuide.rst:
```diff
@@ -51,8 +51,7 @@ another sub-heading (e.g. Powder Diffraction, Algorithms, MSlice), you need to d
 - Add ``New_features`` and ``Bugfixes`` directories within your new directory. The automated script only works with these directories.
-- Add a blank .rst file to each of the ``New_features`` and ``Bugfixes`` directories that has the same name as the directory - this is where the automated script amalgamates the separate release notes.
-- Update the top level release note file with your new heading and sub-headings. For each subheading you need to add an include statement with a link to each new blank .rst file to ensure the notes will be visible for users. For this example it may look something like this
+- Update the top level release note file with your new heading and sub-headings. For each subheading you need to add an ``amalgamate`` statement with a link to each new directory to ensure the notes will be visible for users. The link to the new directory must start with `/` to enable the script to work. For this example it may look something like this
```
We should check for this leading `/` and emit some sort of exception if it's missing, as I can see it being a common mistake

sf1919 replied:
I've incoporated this into a ned method which will generate the path

DavidFair commented on dev-docs/source/Standards/ReleaseNotesGuide.rst:
```diff
@@ -61,24 +60,23 @@ another sub-heading (e.g. Powder Diffraction, Algorithms, MSlice), you need to d
 ------------------------
 
-There are two ways you can preview release notes prior to them being merged in.
+Currently the only are way you can preview release notes prior to them being merged in is as follows:-
```
```suggestion
Currently the only way you can preview release notes prior to them being merged in is as follows:-
```

DavidFair commented on dev-docs/source/Standards/ReleaseNotesGuide.rst:
```diff
@@ -61,24 +60,23 @@ another sub-heading (e.g. Powder Diffraction, Algorithms, MSlice), you need to d
 
-1. Build the documents and view the release note from the build folder as a HTML file. You will need to find it in the directory structure and it may take some time if you have not recently built documentation.
-2. Via the Pull Request (PR). You can view the release note on Github and it will show it using basic .rst rendering. You cannot check all the features you might expect to see when the release note is merged in (e.g. you cannot
+1. Via the Pull Request (PR). You can view the release note on Github and it will show it using basic .rst rendering. You cannot check all the features you might expect to see when the release note is merged in (e.g. you cannot
```
```suggestion
Via a Pull Request (PR): You can view the release note on Github and it will show it using basic .rst rendering. You cannot check all the features you might expect to see when the release note is merged in (e.g. you cannot
```

DavidFair commented on docs/sphinxext/mantiddoc/directives/amalgamate.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        args = self.arguments[0]
+        current_dir = os.path.split(sys.argv[0])[0]
+        path_to_notes = '../../../../../../../../docs/source/' + self.source().rsplit('/', 1)[0] + args
```
Is there a way we can use Git Python to get the root of the repo like this rather than having to use relative paths everywhere like this:
https://github.com/mantidproject/pre-commit-hooks/blob/main/cmake_missing_pytest_files/check_for_missing_py_tests.py#L25

sf1919 committed:
Changes in response to first review

- Removed requirement for developers to add leadinf '/' to amalgamate argument
- created method to create necessary path
- minor update to documentation

DavidFair commented on docs/sphinxext/mantiddoc/directives/amalgamate.py:
```diff
@@ -0,0 +1,67 @@
+        # gets the current directory
+        current_dir = os.getcwd()
+        # the location of documentation
+        source_dir = '../../mantid/docs/source/'
```
This break for me, as it assumes the top level directory is called `mantid`:
```
FileNotFoundError: [Errno 2] No such file or directory: '/home/david/mantid/docs/source/release/v6.4.0/Diffraction/Engineering/New_features'
```

(The above should be mantidpr)

sf1919 committed:
Updated method to get source directory

DavidFair reviewed (approved):
This now works for me and should work with different source folder names
LGTM :+1:
